File: rust/ssot_pipeline/src/degradation.rs
// rust/ssot_pipeline/src/degradation.rs
//
// Execution degradation simulator for the Dual Punch SSOT pipeline.

use crate::audit::audit_two_leg_route_envelope;
use crate::types::ExecutionRunResult;

/// Simple linear congruential generator for reproducible degradation sampling.
/// Used to avoid pulling in an external RNG crate.
pub struct LcgRng {
    state: u64,
}

impl LcgRng {
    /// Create a new RNG with the given seed.
    pub fn new(seed: u64) -> Self {
        Self { state: seed.wrapping_add(1) }
    }

    /// Advance state and return a uniform f64 in [0.0, 1.0).
    pub fn next_f64(&mut self) -> f64 {
        // Knuth multiplicative LCG (64-bit)
        self.state = self.state.wrapping_mul(6_364_136_223_846_793_005).wrapping_add(1_442_695_040_888_963_407);
        (self.state >> 11) as f64 / (1u64 << 53) as f64
    }

    /// Draw from N(mean, std) using Box-Muller transform; clamped to [0, ∞).
    pub fn sample_degradation(&mut self, mean: f64, std: f64) -> f64 {
        let u1 = self.next_f64().max(1e-300);
        let u2 = self.next_f64();
        let z = (-2.0 * u1.ln()).sqrt() * (2.0 * std::f64::consts::PI * u2).cos();
        (mean + std * z).max(0.0)
    }
}

/// Models post-C1 execution variability for a 2-leg constant-product cycle.
pub struct ExecutionDegradationSimulator {
    pub degradation_mean: f64,
    pub degradation_std: f64,
    rng: LcgRng,
}

impl ExecutionDegradationSimulator {
    pub fn new(degradation_mean: f64, degradation_std: f64, seed: u64) -> Self {
        Self {
            degradation_mean,
            degradation_std,
            rng: LcgRng::new(seed),
        }
    }

    fn sample_degradation_factor(&mut self) -> f64 {
        self.rng.sample_degradation(self.degradation_mean, self.degradation_std)
    }

    /// Simulate one execution run with probabilistic profit degradation.
    pub fn simulate_one_run(
        &mut self,
        a_in: f64,
        b_out_1: f64,
        a_out_2: f64,
        p_gross: f64,
        p_net_deterministic: f64,
        c_total: f64,
        _p_fill: f64,
        c2_decision: &str,
        fee1: f64,
        fee2: f64,
    ) -> ExecutionRunResult {
        let audit = audit_two_leg_route_envelope(
            a_in, fee1, b_out_1, b_out_1, fee2, a_out_2, p_gross, p_net_deterministic, c_total, 1e-9,
        );

        if c2_decision != "STRIKE" {
            return ExecutionRunResult {
                a_in,
                b_out_1,
                a_out_2,
                p_gross_deterministic: p_gross,
                p_net_deterministic,
                p_net_actual: 0.0,
                c2_decision: c2_decision.to_string(),
                audit,
            };
        }

        let factor = self.sample_degradation_factor();
        ExecutionRunResult {
            a_in,
            b_out_1,
            a_out_2,
            p_gross_deterministic: p_gross,
            p_net_deterministic,
            p_net_actual: p_net_deterministic * factor,
            c2_decision: c2_decision.to_string(),
            audit,
        }
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn do_nothing_yields_zero_actual_profit() {
        let mut sim = ExecutionDegradationSimulator::new(0.65, 0.35, 42);
        let r = sim.simulate_one_run(1000.0, 995.0, 1008.5, 8.5, 7.5, 1.0, 1.0, "DO_NOTHING", 0.003, 0.0025);
        assert_eq!(r.p_net_actual, 0.0);
        assert_eq!(r.c2_decision, "DO_NOTHING");
    }

    #[test]
    fn strike_produces_nonzero_actual_profit() {
        let mut sim = ExecutionDegradationSimulator::new(0.65, 0.1, 42);
        let r = sim.simulate_one_run(1000.0, 995.0, 1008.5, 8.5, 7.5, 1.0, 1.0, "STRIKE", 0.003, 0.0025);
        assert!(r.p_net_actual >= 0.0, "p_net_actual should be non-negative (clamped)");
        assert_eq!(r.c2_decision, "STRIKE");
    }

    #[test]
    fn audit_is_run_on_each_call() {
        let mut sim = ExecutionDegradationSimulator::new(0.65, 0.35, 0);
        let r = sim.simulate_one_run(1000.0, 995.0, 1008.5, 8.5, 7.5, 1.0, 1.0, "STRIKE", 0.003, 0.0025);
        assert!(r.audit.passed, "Audit should pass for a well-formed plan");
    }
}

File: runtime/pre_test_failure_fix_20260501_214532/python/ssot_pipeline/degradation.py
"""Execution degradation simulator for the Dual Punch SSOT pipeline.

Models post-C1 execution variability without altering the locked inventory
identity (B_in_2 = B_out_1).
"""
from __future__ import annotations

import random
from typing import Optional

from .audit import audit_two_leg_route_envelope
from .types import ExecutionRunResult


class ExecutionDegradationSimulator:
    """Model post-C1 execution variability for a 2-leg constant-product cycle.

    After C1 deterministically computes the optimal size and expected profit,
    real-world execution deviates due to block-inclusion latency, competing
    transactions, and gas-price variance.  This simulator draws a degradation
    factor for each run from a Normal distribution and scales the deterministic
    net profit, without altering the locked inventory identity (B_in_2 = B_out_1).

    The degradation is applied only at the realized-profit scalar level.  The
    AMM math — including the B_in_2 = B_out_1 handoff — is never mutated.

    Parameters
    ----------
    degradation_mean:
        Center of the degradation factor distribution.  A value of 1.0 means
        the expected realized profit equals the deterministic C1 estimate; lower
        values model execution shortfall.  Default is 0.65.
    degradation_std:
        Standard deviation of the degradation factor distribution.  Default
        is 0.35, producing a realistic spread of outcomes that includes both
        sub-zero and near-C1 realizations.
    rng:
        Seeded :class:`random.Random` instance.  Supply a fixed seed for
        reproducible tests; leave ``None`` to use an independent default.
    """

    def __init__(
        self,
        degradation_mean: float = 0.65,
        degradation_std: float = 0.35,
        rng: Optional[random.Random] = None,
    ) -> None:
        self.degradation_mean = float(degradation_mean)
        self.degradation_std = float(degradation_std)
        self._rng = rng if rng is not None else random.Random()

    def _sample_degradation_factor(self) -> float:
        """Draw one degradation factor from N(mean, std), clamped to [0, ∞).

        The distribution is a right-truncated Normal: samples below zero are
        discarded and replaced with 0.0.  When ``degradation_std`` is large
        relative to ``degradation_mean`` (e.g. the default mean=0.65, std=0.35)
        a non-trivial fraction of the theoretical distribution falls below zero;
        clamping shifts the effective mean above ``degradation_mean``.  This
        is intentional: zero-floor realizations model execution runs that fail
        entirely (reverted transactions, MEV displacement) without producing
        negative profit for the strategy.
        """
        return max(0.0, self._rng.gauss(self.degradation_mean, self.degradation_std))

    def simulate_one_run(
        self,
        a_in: float,
        b_out_1: float,
        a_out_2: float,
        p_gross: float,
        p_net_deterministic: float,
        c_total: float,
        p_fill: float,
        c2_decision: str,
        fee1: float = 0.0,
        fee2: float = 0.0,
    ) -> ExecutionRunResult:
        """Simulate one execution run with probabilistic profit degradation.

        When ``c2_decision`` is ``"DO_NOTHING"`` the run contributes zero
        realized profit without any degradation.

        When ``c2_decision`` is ``"STRIKE"`` a degradation factor *f* is drawn
        from ``N(degradation_mean, degradation_std)`` and the actual profit is::

            p_net_actual = p_net_deterministic × f

        The locked inventory identity (B_in_2 = B_out_1) is preserved by always
        passing ``b_out_1`` as ``b_in_2`` to the audit.

        Parameters
        ----------
        a_in:
            C1 optimal input amount (asset A).
        b_out_1:
            Swap 1 output from the deterministic C1 computation.
        a_out_2:
            Swap 2 output from the deterministic C1 computation.
        p_gross:
            Gross profit in asset A (= A_out_2 − A_in).
        p_net_deterministic:
            Net profit in asset A after all costs.
        c_total:
            Owner submission gas component in asset A.
        p_fill:
            Fill probability used to gate the C2 decision; not sampled here.
        c2_decision:
            ``"STRIKE"`` or ``"DO_NOTHING"`` as decided upstream.
        fee1:
            Swap 1 DEX fee rate used in the original C1 computation (decimal).
        fee2:
            Swap 2 DEX fee rate used in the original C1 computation (decimal).

        Returns
        -------
        ExecutionRunResult
        """
        audit = audit_two_leg_route_envelope(
            a_in=a_in,
            fee1=fee1,
            b_out_1=b_out_1,
            b_in_2=b_out_1,   # b_in_2 IS b_out_1 in a correct plan (invariant)
            fee2=fee2,
            a_out_2=a_out_2,
            p_gross=p_gross,
            p_net=p_net_deterministic,
            c_total=c_total,
        )

        if c2_decision != "STRIKE":
            return ExecutionRunResult(
                a_in=a_in,
                b_out_1=b_out_1,
                a_out_2=a_out_2,
                p_gross_deterministic=p_gross,
                p_net_deterministic=p_net_deterministic,
                p_net_actual=0.0,
                c2_decision=c2_decision,
                audit=audit,
            )

        degradation_factor = self._sample_degradation_factor()
        p_net_actual = p_net_deterministic * degradation_factor

        return ExecutionRunResult(
            a_in=a_in,
            b_out_1=b_out_1,
            a_out_2=a_out_2,
            p_gross_deterministic=p_gross,
            p_net_deterministic=p_net_deterministic,
            p_net_actual=p_net_actual,
            c2_decision=c2_decision,
            audit=audit,
        )
